File: pro_realtime_publisher/src/main/java/pro/gmall/realtime/publisher/utils/MyESUtil.java
package pro.gmall.realtime.publisher.utils;

import io.searchbox.client.JestClient;
import io.searchbox.client.JestClientFactory;
import io.searchbox.client.config.HttpClientConfig;
import io.searchbox.core.DocumentResult;
import io.searchbox.core.Get;

import java.io.IOException;

public class MyESUtil {

	JestClientFactory jestClientFactory = null;

	public JestClient getJestClient() {
		if (jestClientFactory == null) {
			build();
		}

		return jestClientFactory.getObject();
	}

	private void build() {
		String esUrl = new MyPropertiesUtil().load("config.properties").getProperty("es.url");

		jestClientFactory = new JestClientFactory();
		jestClientFactory.setHttpClientConfig(new HttpClientConfig.Builder(esUrl)
				.multiThreaded(true)
				.maxTotalConnection(20)
				.connTimeout(10000)
				.readTimeout(1000)
				.build()
		);
	}

	public void closeJestClient(JestClient jestClient) {
		if (jestClient != null) {
			try {
				jestClient.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public static void main(String[] args) {
		JestClient jedisClient = new MyESUtil().getJestClient();
		Get get = new Get.Builder("gmall_dau_info-query", "mid_112").build();
		try {
			DocumentResult execute = jedisClient.execute(get);
			System.out.println("execute = " + execute.getJsonString());
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			new MyESUtil().closeJestClient(jedisClient);
		}
	}
}

File: pro_realtime/src/main/scala/utils/MyEsUtil.scala
package utils

import io.searchbox.client.JestClientFactory
import io.searchbox.client.config.HttpClientConfig
import io.searchbox.core.{Bulk, BulkResult, Index}

import java.util.Properties

object MyEsUtil {

	private val prop: Properties = MyPropertyUtil.load("config.properties")
	private val esUrl: String = prop.getProperty("es.url")

	var jestFactory: JestClientFactory = null

	/**
	 * 获取Jest客户端
	 *
	 * @return
	 */
	def getJsetClient() = {
		if (jestFactory == null) {
			// 创建Jest工厂类方法
			build()
		}
		jestFactory.getObject
	}

	/**
	 * 构建Jest客户端
	 */
	def build() = {
		jestFactory = new JestClientFactory
		jestFactory.setHttpClientConfig(new HttpClientConfig.Builder(esUrl)
		  .multiThreaded(true)
		  .maxTotalConnection(20)
		  .connTimeout(10000)
		  .readTimeout(1000).build())
	}

	/**
	 * 向ES中批量插入数据
	 *
	 * @param infoList  需插入的数据
	 * @param indexName index名称
	 */
	def bulkInsert(infoList: List[(String, Any)], indexName: String): Unit = {
		if (infoList != null && infoList.nonEmpty) {
			val jestClient = getJsetClient()
			val bulkBuilder = new Bulk.Builder()

			// 对批量操作的数据进行遍历
			for ((id, source) <- infoList) {
				val index: Index = new Index.Builder(source)
				  .index(indexName)
				  .id(id)
				  .`type`("_doc")
				  .build()

				bulkBuilder.addAction(index)
			}

			val bulk: Bulk = bulkBuilder.build()
			val bulkResult: BulkResult = jestClient.execute(bulk)

			println("向ES中插入" + bulkResult.getItems.size() + "条数据")
//			println(bulkResult.getJsonString)

			jestClient.close()
		}
	}

}
